core/src/main/java/org/apache/iceberg/PartitionSpecUpdate.java

```diff
@@ -0,0 +1,214 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.Maps;
+import java.util.Map;
+import org.apache.iceberg.transforms.Transform;
+
+/**
+ * PartitionSpec evolution API implementation.
+ */
+class PartitionSpecUpdate implements UpdatePartitionSpec {
+
+  private final TableMetadata base;
+  private final TableOperations ops;
+  private PartitionSpec.Builder newSpecBuilder;
+
+  PartitionSpecUpdate(TableOperations ops) {
+    this.ops = ops;
+    this.base = ops.current();
+    this.newSpecBuilder = PartitionSpec.builderFor(base.schema());
```

Comment:
I don't think this should use the builder to keep track of state. That makes the builder more complicated and leaks concerns from this class to it. This class should be responsible for applying changes and should use the builder to create a new `PartitionSpec` instance without keeping the state from this class in the builder.

I think it makes the most sense to keep track of a list of fields here. The `PartitionField` constructor is accessible, so you can maintain `List<PartitionField>` and make changes to it as update methods are called.

Reply:
@rdblue thanks for the comments. Let me check if this will introduce some duplicate code (same logic in partition spec builder and here).